File: compare/misc/expr.py
#!/usr/bin/env python
# -*- coding: utf-8; -*-

import sys
import operator
import time

sys.setrecursionlimit(1000 * sys.getrecursionlimit())

class ParseError(Exception):
    pass

def expect(tok, toks):
    if accept(tok, toks):
        return toks.pop(0)
    else:
        raise ParseError

def accept(tok, toks):
    return len(toks) and tok == toks[0]

class Node(object):
    __slots__ = ()
    def __repr__(self):
        return str(self.__class__.__name__)
    def evaluate(self):
        raise NotImplementedError("Abstract Base Class")
    def transform(self):
        return self


class Expr(Node):
    op = None

    def __init__(self, l, r=None):
        self.left = l
        self.right = r

    def __repr__(self):
        s = super(Expr, self).__repr__()
        if self.right is None:
            return "".join([s, "(", repr(self.left), ")"])
        else:
            return "".join([s, "(", repr(self.left), ", ", repr(self.right), ")"])

    def __str__(self):
        if self.right is None:
            return "".join(["(", str(self.left), ")"])
        else:
            return "".join(["(", str(self.left), " ", self.op, " ", str(self.right), ")"])

    def transform(self):
        if self.right is None:
            return self.left.transform()
        else:
            return self.__class__(self.left.transform(), self.right.transform())

    @classmethod
    def make(cls, toks):
        toks = lstrip(toks)
        (l, rest) = cls.next.make(toks)
        r = None
        if accept(cls.op, rest):
            rest.pop(0)
            (r, rest) = cls.next.make(lstrip(rest))
        return cls(l, r), rest

class Mul(Expr):
    op = "*"
    def evaluate(self):
        return self.left.evaluate() * self.right.evaluate()

class Add(Expr):
    op = "+"
    def evaluate(self):
        return self.left.evaluate() + self.right.evaluate()

class Num(Expr):

    def transform(self):
        assert self.right is None
        if isinstance(self.left, Expr):
            e = self.left
            assert e.right is None
            return e.left.transform()
        elif isinstance(self.left, NUM):
            return self.left.transform()
        else:
            raise RuntimeError("dont know what to do")

    @classmethod
    def make(cls, toks):
        toks = lstrip(toks)
        if accept("(", toks):
            toks = lstrip(toks)
            expect("(", toks)
            e, rest = Expr.make(lstrip(toks))
            expect(")", rest)
        else:
            lit = ""
            while toks[0].isdigit():
                lit += toks.pop(0)
            e, rest = NUM(int(lit)), lstrip(toks)
        return cls(e), lstrip(rest)

class NUM(Node):
    def __repr__(self):
        s = super(NUM, self).__repr__()
        return "".join([s, "(%i)" % self.value])
    def __str__(self):
        return "%i" % self.value

    def __init__(self, val):
        self.value = val

    def evaluate(self):
        return self.value

    def transform(self):
        return self

Expr.next = Mul
Mul.next = Add
Add.next = Num

def lstrip(toks):
    if len(toks):
        return toks if not toks[0].isspace() else lstrip(toks[1:])
    return toks

def evaluate(ast):
    # print "ast>>", ast
    # print "ast>>", repr(ast)
    return ast.evaluate()

def transform(st):
    # print "st>>", st
    # print "ast>>", repr(st)
    return st.transform()

def parse(string):
    expr, rest = Expr.make(list(string.strip()))
    if len(rest) != 0:
        raise ParseError
    return expr

from random import random, randint, choice

decay = 1.05
def randomExpression(prob):
    p = random()
    if p > prob:
        return NUM(randint(1, 9))
    elif randint(0, 1) == 0:
        return Expr(randomExpression(prob / decay))
    else:
        left = randomExpression(prob / decay)
        op = choice([Mul, Add])
        right = randomExpression(prob / decay)
        return op(left, right)

    
def generate_main(argv):
    size = int(argv[1]) if len(argv) > 1 else 100
    
    r = transform(randomExpression(1))
    while len(str(r)) < size:
        r = transform(Add(r, randomExpression(1)))
    
    print r
    return 0

def main(argv):
    if len(argv) > 1 and argv[1] == "--generate":
        return generate_main(argv[1:])
        
    
    if len(argv) > 1:
        with open(argv[1]) as f:
            inp = f.read()
    else:
        inp = sys.stdin.read()

    ast = parse(inp)
    
    t1 = time.clock()
    st = transform(ast)
    t2 = time.clock()
    
    val = evaluate(st)
#    print val
#    assert val
    
    t = (t2 - t1) * 1000
    print "0:RESULT-cpu:ms: %s\n0:RESULT-total:ms: %s\n0:RESULT-gc:ms: 0.0\n" % (t, t)

    return 0

if __name__ == '__main__':
    try:
        sys.exit(main(sys.argv))
    except SystemExit:
        pass
    except:
        import pdb, traceback
        _type, value, tb = sys.exc_info()
        traceback.print_exception(_type, value, tb)
        pdb.post_mortem(tb)

File: compare/misc/expr.ml
type expr = NUM of int
          | Expr of expr
          | Num of expr
          | Add of expr * expr option
          | Mul of expr * expr option;;
(* Concrete tree parts *)
type cexpr = CNUM of int
           | CAdd of cexpr * cexpr
           | CMul of cexpr * cexpr;;

exception ParseError of char list;;
exception TransformError of expr;;

let rec implode = function
  | [] -> ""
  | x::xs -> (String.make 1 x) ^ (implode xs)

let rec print_tree = function
  | NUM(x)         -> Printf.printf "NUM(%i)" x
  | Num(x)         -> Printf.printf "Num("; print_tree (x); Printf.printf ")";
  | Expr(x)        -> Printf.printf "Expr("; print_tree (x); Printf.printf ")";
  | Add(x, None)   -> Printf.printf "Add("; print_tree (x); Printf.printf ")";
  | Add(x, Some y) -> Printf.printf "Add("; print_tree (x);
                      Printf.printf ", ";   print_tree (y); Printf.printf ")"
  | Mul(x, None)   -> Printf.printf "Mul("; print_tree (x); Printf.printf ")";
  | Mul(x, Some y) -> Printf.printf "Mul("; print_tree (x);
                      Printf.printf ", ";   print_tree (y); Printf.printf ")"

let rec print_ctree = function
  | CNUM(x)        -> Printf.printf "%i" x
  | CAdd(x, y)     -> Printf.printf "("; print_ctree (x);
                      Printf.printf " + ";   print_ctree (y); Printf.printf ")"
  | CMul(x, y)     -> Printf.printf "("; print_ctree (x);
                      Printf.printf " * ";   print_ctree (y); Printf.printf ")"

let relevant_char = function
  | ' ' | '\n' | '\t' -> false
  | _ -> true

let is_digit x = (x >= '0') && (x <= '9')
let int_of_digit x = (int_of_char x) - (int_of_char '0')

let number l =
  let rec loop acc = function
    | x::xs when is_digit x -> loop (10 * acc + (int_of_digit x)) xs
    | l -> ((NUM(acc)), l)
  in
  loop 0 l


let rec num_expr l = match l with
  | x::xs when is_digit x -> let (n, r) = number l in
                             (Num(n), r)
  | '('::xs -> begin let (x, l) = expr xs in
                     match l with
                     | ')'::r -> (Num(x), r)
                     | _ -> raise (ParseError(l))
               end
  | _ -> raise (ParseError(l))
and add_expr l =
  let (x, l) = num_expr l in
  match l with
  | '+'::xs -> let (y, l) = num_expr xs in
               (Add(x, Some y), l)
  | _ -> (Add(x, None), l)
and mul_expr l =
  let (x, l) = add_expr l in
  match l with
  | '*'::xs -> let (y, l) = add_expr xs in
               (Mul(x, Some y), l)
  | _ -> (Mul(x, None), l)
and expr l =
  let (x, r) = mul_expr l in
  (Expr (x), r)


let rec evaluate = function
  | CAdd(x, y) -> evaluate (x) + evaluate (y)
  | CMul(x, y) -> evaluate (x) * evaluate (y)
  | CNUM(x)    -> x

let rec transform = function
  | Num(NUM(x))     -> CNUM(x)
  | Num(Expr(x))    -> transform (x)
  | Expr(x)         -> transform (x)
  | Add(x, None)    -> transform (x)
  | Add(x, Some y)  -> CAdd(transform (x), transform (y))
  | Mul(x, None)    -> transform (x)
  | Mul(x, Some y)  -> CMul(transform (x), transform (y))
  | x               -> raise (TransformError(x))


let parse l =
  let (x, r) = expr l in
  match r with
  | [] -> x
  | _  -> raise (ParseError(r))


let rec read inchar char_list =
  try
    read inchar ((input_char inchar)::char_list)
  with End_of_file -> List.rev char_list

let time f x =
  let tot_start = Unix.gettimeofday () in
  let cpu_start = Sys.time() in
  let res = f x in
  let cpu_stop = Sys.time() in
  let tot_stop = Unix.gettimeofday () in
  let () = Printf.printf "0:RESULT-cpu:ms: %f\n0:RESULT-total:ms: %f\n"
                         ((cpu_stop -. cpu_start) *. 1000.0)
                         ((tot_stop -. tot_start) *. 1000.0)
  in
  res

let _ =
  try
    let chars = match (Array.length Sys.argv) with
      | 1 -> read stdin []
      | _ -> let inchar = open_in Sys.argv.(1) in
             let c = read inchar [] in
             let () = close_in inchar in
             c
    in
    let char_list = List.filter relevant_char chars in
    let st = parse char_list in
    (* let () = print_tree st; Printf.printf "\n" in *)
    let ast = time transform st in
    let () = print_ctree ast; Printf.printf "\n" in
    let value = evaluate ast in
    let () = Printf.printf "%i\n" value in
    0
  with ParseError(l) -> (Printf.printf "Parse error at '%s'\n" (implode l); 1);;
